### PR TITLE
[Feature]: Limit the number of threads for blobnode and datanode to read and write disk

### LINKED ISSUE #1915: [Feature]: Limit the number of threads for blobnode and datanode to read and write disk
**Describe the solution you'd like**
- Limit the number of IO read and write threads of the disk, reduce the high load caused by frequent switching threads of the CPU, and control the read and write speed of the disk at the same time.

**Additional context**
When the load on the disk is relatively high, the go language will automatically generate a large number of threads, and the frequent switching of threads by the CPU will cause a high load, and even more than 10,000 threads may cause the process to panic.

### PR DESCRIPTION
**Pull request by NaturalSelect**
**What this PR does / why we need it**:

Limit the number of threads for blobnode and datanode to read and write disk

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

fixes #1915 

**Special notes for your reviewer**:

~~New dependency:~~
~~* `gopsutil` - In order to get real-time disk load.~~

The implementation of the IO pool takes over:
* `ReadAt`
* `WriteAt`
* `Sync`
* `Fallocate`

~~For datanode, I modified:~~

~~* `Extent.Write`~~
~~* `Extent.Read`~~
~~* `Extent.DeleteTiny`~~
~~* `Extent.TinyExtentRecover`~~
~~* `PersistenceBlockCrc`~~
~~* `PersistenceBaseExtentID`~~
~~* `PreAllocSpaceOnVerifyFile`~~
~~* `loadExtentFromDisk`~~

For blobnode, I modified:

* `blobfile.ReadAt`
* `blobfile.WriteAt`
* `blobfile.Allocate`
* `blobfile.Discard`
* `blobfile.Sync`

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by pengtianyue025** [comment]
@cessory

**Comment by Victor1319** [comment]
here is a merge commit caused by me, which is not suggest.
`[Merge branch 'master' into disk-limit-dev](https://github.com/cubefs/cubefs/pull/1974/commits/329625a997129e437bd65499939fb5232caaaeb8)`
please rebase master branch into your local branch, and then try to make a force push again.

**Comment by cessory** [comment]
It is recommended to submit a small number of changes at a time

**Comment by NaturalSelect** [comment]
> here is a merge commit caused by me, which is not suggest. `[Merge branch 'master' into disk-limit-dev](https://github.com/cubefs/cubefs/pull/1974/commits/329625a997129e437bd65499939fb5232caaaeb8)` please rebase master branch into your local branch, and then try to make a force push again.

Ok

**Comment by NaturalSelect** [comment]
> It is recommended to submit a small number of changes at a time

Ok, I split it into several commits.

**Comment by cessory** [comment]
> > It is recommended to submit a small number of changes at a time
> 
> Ok, I split it into several commits.

I'll close this pr

**Review by mawei029 — COMMENTED** [review]
增加单测，增加conf配置，代码风格

File: blobstore/util/iopool/io_scheduler.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+		pool.Run(func() {
+			cb.runLoop()
+		})
+	}
+}
+
+type SimpleIoScheduler struct {
+	pool            taskpool.TaskPool
+	controllerTable map[uint64]*FileController
+	tabelLock       *sync.RWMutex
```
[the remark this answers is not in the document]
Comment (NaturalSelect) [reply blobstore/util/iopool/io_scheduler.go]: fixed!

File: blobstore/util/iopool/io_task.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+func (task *IoTask) GetOffset() uint64 {
+	return task.offset
+}
+
+func (task *IoTask) GetSize() uint64 {
+	return uint64(len(task.date))
+}
+
+func (task *IoTask) Vailed() bool {
```
[the remark this answers is not in the document]
Comment (NaturalSelect) [reply blobstore/util/iopool/io_task.go]: fixed!

File: blobstore/util/iopool/io_task.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	WriteAt(b []byte, off int64) (n int, err error)
+	Sync() error
+	Fd() uintptr
+}
+
+type IoTask struct {
+	taskType   IoTaskType
+	handleID   uint64
+	handle     FileHandle
+	date       []byte
```
[the remark this answers is not in the document]
Comment (NaturalSelect) [reply blobstore/util/iopool/io_task.go]: fixed!

File: blobstore/util/iopool/io_task.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+func NewAllocIoTask(handle FileHandle, id uint64, mode uint32, offset uint64, size uint64) *IoTask {
+	task := newIoTask(AllocIoTask, handle, id, offset, nil, false)
+	task.allocSize = size
+	task.allocMode = mode
+	return task
+}
+
+func (task *IoTask) WaitAndClose() (int, error) {
```
[the remark this answers is not in the document]
Comment (NaturalSelect) [reply blobstore/util/iopool/io_task.go]: fixed!

File: blobstore/util/iopool/io_scheduler.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
+// Copyright 2023 The CubeFS Authors.
```
[the remark this answers is not in the document]
Comment (NaturalSelect) [reply blobstore/util/iopool/io_scheduler.go:1]: 已添加测试

File: blobstore/util/iopool/io_scheduler.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+func (scheduler *ShardedIoScheduler) Schedule(task *IoTask) {
+	index := uint(task.GetHandleID() % uint64(len(scheduler.subScheduler)))
+	scheduler.subScheduler[index].Schedule(task)
+}
+
+func NewShardedIoScheduler(count uint32, pool taskpool.TaskPool) *ShardedIoScheduler {
+	if count == 0 {
+		count = 64
```
[the remark this answers is not in the document]
Comment (NaturalSelect) [reply blobstore/util/iopool/io_scheduler.go]: fixed!

File: blobstore/util/iopool/io_task.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+func newIoTask(taskType IoTaskType, handle FileHandle, handleID uint64, offset uint64, data []byte, isSync bool) *IoTask {
+	return &IoTask{
+		taskType:   taskType,
+		handleID:   handleID,
+		handle:     handle,
+		date:       data,
+		offset:     offset,
+		resultChan: make(chan interface{}, 1),
+		isSync:     isSync,
+		allocMode:  0,
```
[the remark this answers is not in the document]
Comment (NaturalSelect) [reply blobstore/util/iopool/io_task.go]: fixed!

File: blobstore/blobnode/core/disk/disk.go [hunk 8]
```diff
@@ -487,6 +500,18 @@ func newDiskStorage(ctx context.Context, conf core.Config) (ds *DiskStorage, err
...
+	// setting pools
+	if conf.WriteThreadCnt == 0 {
+		conf.WriteThreadCnt = 2
+	}
+	if conf.ReadThreadCnt == 0 {
+		conf.ReadThreadCnt = 4
+	}
+	writePool := taskpool.New(int(conf.WriteThreadCnt), int(conf.WriteThreadCnt))
+	readPool := taskpool.New(int(conf.ReadThreadCnt), int(conf.ReadThreadCnt))
+	writeScheduler := iopool.NewShardedIoScheduler(1024, writePool)
```
[the remark this answers is not in the document]
Comment (NaturalSelect) [reply blobstore/blobnode/core/disk/disk.go]: 用户可能很难正常配置此项目，已修改为const

File: blobstore/blobnode/core/disk/disk.go [hunk 9]
```diff
@@ -487,6 +500,18 @@ func newDiskStorage(ctx context.Context, conf core.Config) (ds *DiskStorage, err
 		return nil, err
 	}
 
+	// setting pools
+	if conf.WriteThreadCnt == 0 {
+		conf.WriteThreadCnt = 2
```
[the remark this answers is not in the document]
Comment (NaturalSelect) [reply blobstore/blobnode/core/disk/disk.go:511]: 用户可能在配置文件中，将此项修改为`0`，在此处的`if`的必要的。

[the remark this answers is not in the document]
Comment (NaturalSelect) [reply blobstore/blobnode/core/disk/disk.go]: fixed!

File: blobstore/util/iopool/io_scheduler.go [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+				return cb.tasks[i].offset < cb.tasks[j].offset
+			})
+			isSync := false
+			for _, task := range cb.tasks {
+				task.Exec()
+				isSync = isSync || task.IsSync()
+			}
+			// sync file
+			if isSync {
+				cb.tasks[0].Sync()
```
[the remark this answers is not in the document]
Comment (NaturalSelect) [reply blobstore/util/iopool/io_scheduler.go:79]: 多个task指向同一个handle，因此只需要对一个task调用一次sync

File: blobstore/util/iopool/io_scheduler.go [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+)
+
+type IoScheduler interface {
+	Schedule(task *IoTask)
+}
+
+type FileController struct {
+	taskCount    uint32
+	pendingTasks []*IoTask
+	tasks        []*IoTask
```
Comment (mawei029) [inline blobstore/util/iopool/io_scheduler.go:32]: 这里是否可以优化一下，改成长度限制的数组/slice/chan之类的，以防内存占用太多。把长度放到conf里可配。

File: blobstore/util/iopool/io_scheduler.go [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+	"github.com/cubefs/cubefs/blobstore/util/taskpool"
+)
+
+type IoScheduler interface {
+	Schedule(task *IoTask)
+}
+
+type FileController struct {
+	taskCount    uint32
+	pendingTasks []*IoTask
```
Comment (mawei029) [inline blobstore/util/iopool/io_scheduler.go:31]: 是否可以增加个高低优先级队列。
1、目前所有io和请求都入的同一个队列，没区分优先级。长期考虑来看，可以让正常业务进入高优先级队列，后台一些修补任务之类的IO入队低优先级别队列；
2、或者只用一个队列，但是后台/低优先级任务，占用队列最多10%之类，把队列更多空间留给在线实时高优先级任务。

File: blobstore/util/iopool/io_scheduler.go [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+	}
+}
+
+func (cb *FileController) Submit(task *IoTask, pool taskpool.TaskPool) {
+	count := cb.addTaskCount(1)
+	cb.addTask(task)
+	if count == 1 {
+		// if count is 1, we need to submit a task
+		pool.Run(func() {
+			cb.runLoop()
```
Comment (mawei029) [inline blobstore/util/iopool/io_scheduler.go:103]: 确认下大压力高并发场景是否会发生饿死，是通过配置来解决吗，pool协程个数等于subScheduler长度？还是其它办法？
比如这里const 1024大小的subScheduler，在subScheduler里元素[0]，如果一直有新的task请求，cb.runLoop()一直不退出，确认下是否会导致其它1023个scheduler饿死？

**Review by mawei029 — COMMENTED** [review]
一些流程逻辑上的优化点

File: blobstore/util/iopool/io_scheduler.go (hunk 11 above)
Comment (NaturalSelect) [reply blobstore/util/iopool/io_scheduler.go:32]: 上层应该有limit

File: util/chanutil/queue.go [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+				batch = append(batch, item)
+			default:
+				stop = true
+			}
+		}
+	}
+	return batch
+}
+
+func (q Queue) DequeAll() []interface{} {
```
Comment (mawei029) [inline util/chanutil/queue.go:82]: 这两个函数可以优化一下，DequeAll和DequeBatch大部分一样，复用。

File: blobstore/util/iopool/shared_io_scheduler.go [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+type SharedIoScheduler struct {
+	queue chanutil.Queue
+	pool  taskpool.TaskPool
+}
+
+func NewSharedIoScheduler(workerCount int, queueDepth int) *SharedIoScheduler {
+	queue := chanutil.NewQueue(queueDepth)
+	pool := taskpool.New(workerCount, 0)
+	scheduler := &SharedIoScheduler{
```
Comment (mawei029) [inline blobstore/util/iopool/shared_io_scheduler.go:32]: 可以更精简一点。	
scheduler := &SharedIoScheduler{
		queue: chanutil.NewQueue(queueDepth),
		pool:  taskpool.New(workerCount, 0),
}

File: blobstore/util/iopool/shared_io_scheduler.go [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+)
+
+type SharedIoScheduler struct {
+	queue chanutil.Queue
+	pool  taskpool.TaskPool
+}
+
+func NewSharedIoScheduler(workerCount int, queueDepth int) *SharedIoScheduler {
+	queue := chanutil.NewQueue(queueDepth)
+	pool := taskpool.New(workerCount, 0)
```
Comment (mawei029) [inline blobstore/util/iopool/shared_io_scheduler.go:31]: taskpool.New(workerCount, workerCount)
本文件里，第二个参数为0不合理。queue不做Close，43行pool.Run里面的for不会退出，一直执行。其它workerCount没机会调用，没能体现出来workerCount

File: blobstore/util/iopool/partition_io_scheduler.go [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+	}
+}
+
+func NewPartitionIoScheduler(workerCount int, queueDepth int) *PartitionIoScheduler {
+	queues := make([]chanutil.Queue, 0, workerCount)
+	for i := 0; i < workerCount; i++ {
+		queue := chanutil.NewQueue(queueDepth)
+		queues = append(queues, queue)
+	}
+	pool := taskpool.New(workerCount, 0)
```
Comment (mawei029) [inline blobstore/util/iopool/partition_io_scheduler.go:77]: taskpool.New(workerCount, workerCount)

File: blobstore/util/iopool/partition_io_scheduler.go [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+	scheduler := &PartitionIoScheduler{
+		queues: queues,
+		pool:   pool,
+	}
+	scheduler.startWorkers(workerCount, queueDepth)
+	return scheduler
+}
+
+func (s *PartitionIoScheduler) getQueue(task *IoTask) chanutil.Queue {
+	index := int(task.GetHandleID() % uint64(len(s.queues)) % math.MaxInt)
```
Comment (mawei029) [inline blobstore/util/iopool/partition_io_scheduler.go:87]: 是否有必要取模两次，index := task.GetHandleID() % uint64(len(s.queues)) 是否也可以？

File: util/chanutil/queue.go [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+		return false
+	}
+}
+
+func (q Queue) Deque() (interface{}, bool) {
+	item, ok := <-q.queue
+	return item, ok
+}
+
+func (q Queue) DequeBatch(maxCount int) []interface{} {
```
Comment (mawei029) [inline util/chanutil/queue.go:60]: 可以适当在函数前面加点注释说明作用。
看代码本意是想在没有close的情况下，至少取出来一个，最多取出来maxCount个。

Comment (NaturalSelect) [reply util/chanutil/queue.go]: 已添加注释

File: blobstore/util/iopool/partition_io_scheduler.go (hunk 18 above)
Comment (NaturalSelect) [reply blobstore/util/iopool/partition_io_scheduler.go]: 已修改

File: blobstore/util/iopool/partition_io_scheduler.go (hunk 17 above)
Comment (NaturalSelect) [reply blobstore/util/iopool/partition_io_scheduler.go]: review完成之后会合并

**Comment by NaturalSelect** [comment]
Please review again.

**Comment by mawei029** [comment]
不合入master，这个特性先合入blobstore-develop分支。https://github.com/cubefs/cubefs/tree/blobstore-develop

**Comment by NaturalSelect** [comment]
> 不合入master，这个特性先合入blobstore-develop分支。https://github.com/cubefs/cubefs/tree/blobstore-develop

@mawei029 OK

**Review by cessory — APPROVED** [review]

**Pull request merged**